esphome/components/esp32/__init__.py

```diff
@@ -386,7 +397,7 @@ def _check_versions(value):
         value[CONF_SOURCE] = value.get(
             CONF_SOURCE, _format_framework_arduino_version(version)
         )
-        if value[CONF_SOURCE].startswith("http"):
+        if _is_framework_url(value[CONF_SOURCE]):
```

Comment:
```suggestion
        if "@" not in value[CONF_SOURCE]:
```
It is simpler and probably better since using @ might be a valid url as it used for user/password info. Thoughts?

Reply:
For consistency, I think it would be better to accept sources that contain '@' instead of rejecting them as suggested.

According to the [PlatformIO package specifications](https://docs.platformio.org/en/latest/core/userguide/pkg/cmd_install.html#package-specifications), '@' is a valid character within a repository URL and it shows examples.

I documented the `source` configuration variable as a URL since that's what it's mostly useful for so my patch does not handle the case where the source is "owner/package@version" to look up in the PlatformIO registry.  The `version` configuration variable already provides better solution for specifying a well-known version.

- `_is_framework_url("scheme://test")` returns True
- `_is_framework_url("scheme://foo@bar/test")` returns True
- `_is_framework_url("owner/package@1.2.3")` returns False
- `_is_framework_url("owner/package@scheme://test/")` returns False

What do you think?